snap/info_snap_yaml.go

```diff
@@ -21,18 +21,198 @@ package snap
+	Attrs     map[string]interface{} `yaml:"attrs,omitempty"`
+	Apps      []string               `yaml:"apps,omitempty"`
+	Label     string                 `yaml:"label"`
+}
+
+type slotYaml struct {
+	Interface string                 `yaml:"interface"`
+	Attrs     map[string]interface{} `yaml:"attrs,omitempty"`
+	Apps      []string               `yaml:"apps,omitempty"`
+	Label     string                 `yaml:"label"`
+}
+
+type appYaml struct {
+	Command   string   `yaml:"command"`
+	SlotNames []string `yaml:"slots,omitempty"`
+	PlugNames []string `yaml:"plugs,omitempty"`
+}
+
 // InfoFromSnapYaml creates a new info based on the given snap.yaml data
 func InfoFromSnapYaml(yamlData []byte) (*Info, error) {
-	var s Info
-	err := yaml.Unmarshal(yamlData, &s)
+	var y snapYaml
+	err := yaml.Unmarshal(yamlData, &y)
 	if err != nil {
 		return nil, fmt.Errorf("info failed to parse: %s", err)
 	}
-
+	// Construct snap skeleton, without apps, plugs and slots
+	snap := &Info{
+		Name:        y.Name,
+		Developer:   y.Developer,
+		Version:     y.Version,
+		Type:        y.Type,
+		Channel:     y.Channel,
+		Description: y.Description,
+		Apps:        make(map[string]*AppInfo),
+		Plugs:       make(map[string]*PlugInfo),
+		Slots:       make(map[string]*SlotInfo),
+	}
+	// Collect top-level definitions of plugs
+	for name, data := range y.Plugs {
+		iface, attrs, err := convertToSlotOrPlugData("plug", name, data)
+		if err != nil {
+			return nil, err
+		}
+		snap.Plugs[name] = &PlugInfo{
+			Snap:      snap,
+			Name:      name,
+			Interface: iface,
+			Attrs:     attrs,
+			Apps:      make(map[string]*AppInfo),
+		}
+	}
+	// Collect top-level definitions of slots
+	for name, data := range y.Slots {
+		iface, attrs, err := convertToSlotOrPlugData("slot", name, data)
+		if err != nil {
+			return nil, err
+		}
+		snap.Slots[name] = &SlotInfo{
+			Snap:      snap,
+			Name:      name,
+			Interface: iface,
+			Attrs:     attrs,
+			Apps:      make(map[string]*AppInfo),
+		}
+	}
+	// Collect app-level implicit definitions of plugs and slots
+	for _, yApp := range y.Apps {
+		for _, plugName := range yApp.PlugNames {
+			if _, ok := snap.Plugs[plugName]; !ok {
+				snap.Plugs[plugName] = &PlugInfo{
+					Snap:      snap,
+					Name:      plugName,
+					Interface: plugName,
+					Apps:      make(map[string]*AppInfo),
+				}
+			}
+		}
+		for _, slotName := range yApp.SlotNames {
+			if _, ok := snap.Slots[slotName]; !ok {
+				snap.Slots[slotName] = &SlotInfo{
+					Snap:      snap,
+					Name:      slotName,
+					Interface: slotName,
+					Apps:      make(map[string]*AppInfo),
+				}
+			}
+		}
+	}
+	// Collect definitions of apps
+	for appName, yApp := range y.Apps {
+		snap.Apps[appName] = &AppInfo{
+			Snap:    snap,
+			Name:    appName,
+			Command: yApp.Command,
+			Plugs:   make(map[string]*PlugInfo),
+			Slots:   make(map[string]*SlotInfo),
+		}
+	}
+	// Remember which plugs and slots are app-bound
+	appBoundPlug := make(map[string]bool)
```

Comment:
No need for that. After the loop above, Iterate over plugs. If len(plug.Apps) == 0, it's not app bound. Same for slots.